cpp/src/io/csv/csv_reader.cu

```diff
@@ -197,6 +198,84 @@ __device__ void setBit(gdf_valid_type* address, int bit) {
 		return base_padding + num_columns * column_bytes; 
 	}
 }
+/**
+* @brief Removes the first and Last quote in the string
+*/
+string removeQoutes(string str, char quotechar) {
+	// Exclude first and last quotation char
+	const size_t first_quote = str.find(quotechar);
+	if (first_quote != string::npos) {
+		str.erase(first_quote, 1);
+	}
+	const size_t  last_quote = str.rfind(quotechar);
+	if (last_quote != string::npos) {
+		str.erase(last_quote, 1);
+	}
+
+	return str;
+}
+
+/**
+ * @brief Parse the first row to set the column names in the raw_csv parameter 
+ *
+ * The first row can be either the header row, or the first data row
+ *
+ * @param[in,out] raw_csv Structure containing the csv parsing parameters
+ * and intermediate results
+ * 
+ * @return gdf_error with error code on failure, otherwise GDF_SUCCESS
+*/
+gdf_error setColumnNamesFromCsv(raw_csv_t* raw_csv) {
+	vector<char> first_row = raw_csv->header;
+	// No header, read the first data row
+	if (first_row.empty()) {
+		cu_recstart_t first_row_len{};
+		// If file only contains one row, raw_csv->recStart[1] is not valid
+		if (raw_csv->num_records > 1) {
+			CUDA_TRY(cudaMemcpy(&first_row_len, raw_csv->recStart + 1, sizeof(cu_recstart_t), cudaMemcpyDefault));
+		}
+		else {
+			// File has one row - use the file size for the row size
+			first_row_len = raw_csv->num_bytes / sizeof(char);
+		}
+		first_row.resize(first_row_len);
+		CUDA_TRY(cudaMemcpy(first_row.data(), raw_csv->data, raw_csv->num_bytes, cudaMemcpyDefault));
+	}
+
+	int num_cols = 0;
+
+	bool quotation	= false;
+	for (size_t pos = 0, prev = 0; pos < first_row.size(); ++pos) {
+		// Flip the quotation flag if current character is a quotechar
+		if(first_row[pos] == raw_csv->opts.quotechar) {
+			quotation = !quotation;
+		}
+		else if (!quotation &&
+				 (first_row[pos] == raw_csv->opts.delimiter ||
+				 first_row[pos] == raw_csv->opts.terminator)) {
+			// Got to the end of a column
+			if (raw_csv->header_row >= 0) {
+				// first_row is the header, add the column name
+				string new_col_name(first_row.data() + prev, pos - prev);
+				raw_csv->col_names.push_back(removeQoutes(new_col_name, raw_csv->opts.quotechar));
+			}
+			else {
+				// first_row is the first data row, add the automatically generated name
+				raw_csv->col_names.push_back(raw_csv->prefix + std::to_string(num_cols));
+			}
+			num_cols++;
+
+			// Skip adjacent delimiters if delim_whitespace is set
+			while (raw_csv->opts.multi_delimiter && 
+				   first_row[pos] == raw_csv->opts.delimiter && 
+				   first_row[pos + 1] == raw_csv->opts.delimiter) {
```

Comment:
Do we need to add a test if `pos < first_row.size()`?
Consider a CSV file consisting of one data row only and does not have a line terminator.

Reply:
fixed in the latest commit